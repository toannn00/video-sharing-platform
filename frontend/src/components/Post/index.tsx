import { Form, Button, Input } from "antd";
import {
  ArrowLeftOutlined,
  CloudUploadOutlined,
  ReloadOutlined,
} from "@ant-design/icons";
import { useNavigate } from "react-router";
import { useState } from "react";
import videoService from "../../services/video.service";
import Video from "../../types/video.type";

export const Post = () => {
  const [form] = Form.useForm();
  const navigate = useNavigate();
  const [token] = useState(localStorage.getItem("token") || "");
  const [loading, setLoading] = useState(false);

  const onFinish = async (values: Video) => {
    const result = await videoService.create(values, token);

    if (result) {
      form.resetFields();
      setLoading(true);
      back();
    } else {
      setLoading(false);
    }
  };

  const back = () => {
    navigate("/");
  };

  const formClear = () => {
    form.resetFields();
  };

  return (
    <Form
      form={form}
      name="post-form"
      labelCol={{ span: 8 }}
      wrapperCol={{ span: 16 }}
      style={{ maxWidth: 600 }}
      initialValues={{ remember: true }}
      onFinish={onFinish}
      labelAlign="left"
    >
      <Form.Item
        label="Video Title"
        name="title"
        rules={[{ required: true, message: "Input a video title" }]}
      >
        <Input />
      </Form.Item>
      <Form.Item
        label="Youtube URL"
        name="url"
        rules={[
          { required: true, message: "Input valid YouTube URL" },
          {
            pattern: new RegExp("^(https?://)?(www.youtube.com|youtu.be)/.+$"),
            message: "Input a valid YouTube URL",
          },
        ]}
      >
        <Input />
      </Form.Item>
      <Form.Item
        label="Description"
        name="description"
        rules={[{ required: true, message: "Input valid YouTube description" }]}
      >
        <Input.TextArea />
      </Form.Item>

      <Form.Item wrapperCol={{ span: 24 }}>
        <div style={{ display: "flex", gap: 10, justifyContent: "flex-end" }}>
          <Button danger onClick={back}>
            <ArrowLeftOutlined /> Back
          </Button>
          <Button onClick={formClear}>
            <ReloadOutlined /> Clear
          </Button>
          <Button
            style={{ backgroundColor: "black" }}
            htmlType="submit"
            type="primary"
            loading={loading}
          >
            <CloudUploadOutlined />
            Post
          </Button>
        </div>
      </Form.Item>
    </Form>
  );
};

export default Post;
