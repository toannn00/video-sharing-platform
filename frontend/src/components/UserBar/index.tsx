import { Button } from "antd";
import User from "../../types/user.type";
import { useNavigate } from "react-router-dom";

const UserBar = ({ email }: Partial<User>) => {
  const navigate = useNavigate();

  return (
    <div
      style={{
        display: "flex",
        alignItems: "center",
        gap: 10,
        margin: "0 20px",
      }}
    >
      <div>
        Welcome <b>{email}</b>{" "}
      </div>
      <Button onClick={() => navigate("/post")}>Post</Button>
    </div>
  );
};

export default UserBar;
