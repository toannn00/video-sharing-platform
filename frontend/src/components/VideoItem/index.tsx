import { lazy, Suspense } from "react";
import Video from "../../types/video.type";
import { Row, Col } from "antd";
import { useMediaQuery } from "react-responsive";
import { useInView } from "react-intersection-observer";

const ReactPlayer = lazy(() => import("react-player"));

export const VideoItem = ({ video }: { video: Video }) => {
  const { title, description, url, email } = video;
  const isMobile = useMediaQuery({ maxWidth: 1280 });
  const { ref, inView } = useInView({
    threshold: 0.1,
    triggerOnce: true,
  });

  const truncate = (text: string) => {
    return text.length > 200 ? text.substring(0, 200) + "..." : text;
  };

  const renderPlayer = () => {
    if (!inView) return <div style={{ height: 240, background: "#f0f0f0" }} />;

    return (
      <Suspense
        fallback={<div style={{ height: 240, background: "#f0f0f0" }} />}
      >
        <ReactPlayer
          width={isMobile ? "100%" : 426}
          height={240}
          url={url}
          light={true}
          playing={false}
        />
      </Suspense>
    );
  };

  if (isMobile) {
    return (
      <Row gutter={10} style={{ padding: 16 }} data-testid="movie-item">
        <Col span={24}>
          <div ref={ref} style={{ textAlign: "center", marginBottom: 16 }}>
            {renderPlayer()}
          </div>
        </Col>
        <Col span={24}>
          <div style={{ textAlign: "left" }}>
            <h2 style={{ margin: "0 0 8px" }}>{title}</h2>
            <p style={{ margin: "0 0 8px" }}>
              <b>Uploaded by:</b> {email}
            </p>
            <p style={{ margin: 0 }}>
              <b>Description:</b>
            </p>
            <p style={{ margin: 0 }}>{truncate(description)}</p>
          </div>
        </Col>
      </Row>
    );
  }

  return (
    <Row gutter={10} style={{ padding: 16 }} data-testid="movie-item">
      <Col span={10}>
        <div ref={ref} style={{ textAlign: "right" }}>
          {renderPlayer()}
        </div>
      </Col>
      <Col span={10} offset={3}>
        <div style={{ textAlign: "left" }}>
          <h2 style={{ margin: "0 0 8px" }}>{title}</h2>
          <p style={{ margin: "0 0 8px" }}>
            <b>Uploaded by:</b> {email}
          </p>
          <p style={{ margin: 0 }}>
            <b>Description:</b>
          </p>
          <p style={{ margin: 0 }}>{truncate(description)}</p>
        </div>
      </Col>
    </Row>
  );
};

export default VideoItem;
