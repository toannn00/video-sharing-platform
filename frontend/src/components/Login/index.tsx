import { Form, Input, Button, Row, Col } from "antd";
import UserBar from "../UserBar";
import User from "../../types/user.type";
import { useState } from "react";
import authService from "../../services/auth.service";
import { useNavigate } from "react-router-dom";

export const Login = () => {
  const [email, setEmail] = useState(localStorage.getItem("email") || "");
  const navigate = useNavigate();

  const handleLogout = () => {
    authService.logout();
    setEmail("");
    navigate("/");
  };

  const handleLogin = async (values: User) => {
    const res = await authService.login({
      email: values.email,
      password: values.password,
    });

    if (res) {
      setEmail(values.email);
    }
  };

  const onFinish = (values: User) => {
    handleLogin(values);
  };

  return email ? (
    <div style={{ display: "flex" }}>
      <UserBar email={email} />
      <Button danger onClick={handleLogout}>
        Logout
      </Button>
    </div>
  ) : (
    <Form
      name="login-form"
      wrapperCol={{ span: 24 }}
      style={{ maxWidth: 600 }}
      initialValues={{ remember: true }}
      onFinish={onFinish}
      autoComplete="off"
    >
      <Row gutter={8}>
        <Col span={9}>
          <Form.Item
            name="email"
            rules={[{ required: true, message: "Enter a valid email" }]}
            style={{ marginBottom: 0 }}
          >
            <Input placeholder="Email" type="email" />
          </Form.Item>
        </Col>
        <Col span={9}>
          <Form.Item
            name="password"
            rules={[
              {
                required: true,
                message: "Enter at least 8 characters",
                min: 8,
              },
            ]}
            style={{ marginBottom: 0 }}
          >
            <Input.Password placeholder="Password" />
          </Form.Item>
        </Col>
        <Col span={6}>
          <Form.Item wrapperCol={{ span: 24 }} style={{ marginBottom: 0 }}>
            <Button
              style={{ backgroundColor: "black" }}
              type="primary"
              htmlType="submit"
            >
              Login/Register
            </Button>
          </Form.Item>
        </Col>
      </Row>
    </Form>
  );
};
