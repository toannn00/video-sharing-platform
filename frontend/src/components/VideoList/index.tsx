import { useMemo } from "react";
import Video from "../../types/video.type";
import { Empty } from "antd";
import VideoItem from "../VideoItem";
import { useMediaQuery } from "react-responsive";

export const VideoList = ({ videoList }: { videoList: Video[] }) => {
  const isMobile = useMediaQuery({ maxWidth: 1280 });

  const containerStyle = useMemo(
    () => ({
      width: isMobile ? "100%" : "70%",
      margin: isMobile ? "20px auto" : "40px auto 0",
      padding: isMobile ? 10 : 30,
    }),
    [isMobile]
  );

  return (
    <div style={containerStyle}>
      {videoList.length > 0 ? (
        videoList.map((video, index) => (
          <VideoItem key={`${video.url}-${index}`} video={video} />
        ))
      ) : (
        <Empty />
      )}
    </div>
  );
};

export default VideoList;
