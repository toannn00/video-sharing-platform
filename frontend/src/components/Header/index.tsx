import { Row, Col, Divider } from "antd";
import { Login } from "../Login";
import { Link } from "react-router-dom";

export const Header = () => {
  return (
    <>
      <Row align="middle">
        <Col span={9} offset={3}>
          <Link to="/">
            <div style={{ display: "flex" }}>
              <h1>Video App</h1>
            </div>
          </Link>
        </Col>
        <Col span={9} offset={1}>
          <Login />
        </Col>
      </Row>
      <Divider style={{ marginTop: 5 }} />
    </>
  );
};

export default Header;
