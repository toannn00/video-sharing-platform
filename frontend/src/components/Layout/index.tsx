import Header from "../Header";

export const Layout = ({ children }: { children: React.ReactNode }) => (
  <>
    <Header />
    {children}
  </>
);

export default Layout;
