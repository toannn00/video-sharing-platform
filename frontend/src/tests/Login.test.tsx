import "@testing-library/jest-dom";
import ShallowRenderer from "react-test-renderer/shallow";
import { Login } from "../components/Login";
import { vi } from "vitest";

vi.mock("react-router-dom", () => ({
  useNavigate: () => vi.fn(),
}));

describe("Login", () => {
  test("should render and match the snapshot", () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(<Login />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
