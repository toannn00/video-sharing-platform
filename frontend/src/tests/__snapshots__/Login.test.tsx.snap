// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > should render and match the snapshot 1`] = `
<Form
  autoComplete="off"
  initialValues={
    {
      "remember": true,
    }
  }
  name="login-form"
  onFinish={[Function]}
  style={
    {
      "maxWidth": 600,
    }
  }
  wrapperCol={
    {
      "span": 24,
    }
  }
>
  <Row
    gutter={8}
  >
    <Col
      span={9}
    >
      <InternalFormItem
        name="email"
        rules={
          [
            {
              "message": "Enter a valid email",
              "required": true,
            },
          ]
        }
        style={
          {
            "marginBottom": 0,
          }
        }
      >
        <Input
          placeholder="Email"
          type="email"
        />
      </InternalFormItem>
    </Col>
    <Col
      span={9}
    >
      <InternalFormItem
        name="password"
        rules={
          [
            {
              "message": "Enter at least 8 characters",
              "min": 8,
              "required": true,
            },
          ]
        }
        style={
          {
            "marginBottom": 0,
          }
        }
      >
        <Input.Password
          placeholder="Password"
        />
      </InternalFormItem>
    </Col>
    <Col
      span={6}
    >
      <InternalFormItem
        style={
          {
            "marginBottom": 0,
          }
        }
        wrapperCol={
          {
            "span": 24,
          }
        }
      >
        <Button
          htmlType="submit"
          style={
            {
              "backgroundColor": "black",
            }
          }
          type="primary"
        >
          Login/Register
        </Button>
      </InternalFormItem>
    </Col>
  </Row>
</Form>
`;
